hazelcast/src/main/java/com/hazelcast/client/impl/ClientEndpointImpl.java

```diff
@@ -198,6 +202,10 @@ public void addListenerDestroyAction(final String service, final String topic, f
     @Override
     public void addDestroyAction(UUID registrationId, Callable<Boolean> removeAction) {
         removeListenerActions.put(registrationId, removeAction);
+        if (destroyed) {
```

Comment:
I think the following would also fix the leak and will not call the `removeAction` twice in any code path

```java
    @Override
    public void addDestroyAction(UUID registrationId, Callable<Boolean> removeAction) {
        synchronized (removeListenerActions) {
            if (destroyed) {
                callRemoveAction(removeAction);
            } else {
                removeListenerActions.put(registrationId, removeAction);
            }
        }
    }
```

Reply:
Calling the remove action is doing an invocation and waits for the results. We don't want to create a potential deadlock. Since registration id's are unique, calling the remove action twice looks ok to me.

Reply:
What about the following then ?
```java
    @Override
    public void addDestroyAction(UUID registrationId, Callable<Boolean> removeAction) {
        synchronized (removeListenerActions) {
            if (!destroyed) {
                removeListenerActions.put(registrationId, removeAction);
                return;
            }
        }
        callRemoveAction(removeAction);
    }
```